lib/grape/api.rb

```diff
@@ -155,18 +168,21 @@ def add_head_not_allowed_methods_and_options_methods
             allow_header = (self.class.namespace_inheritable(:do_not_route_options) ? allowed_methods : [Grape::Http::Headers::OPTIONS] | allowed_methods).join(', ')
 
             unless self.class.namespace_inheritable(:do_not_route_options)
-              generate_options_method(path, allow_header) unless allowed_methods.include?(Grape::Http::Headers::OPTIONS)
+              generate_options_method(path, allow_header, config[:requirements]) unless allowed_methods.include?(Grape::Http::Headers::OPTIONS)
             end
 
-            generate_not_allowed_method(path, allowed_methods, allow_header)
+            generate_not_allowed_method(regexp, allowed_methods, allow_header, config[:requirements], config[:endpoint])
+            self.class.namespace_stackable(:format, [])
           end
         end
       end
     end
 
     # Generate an 'OPTIONS' route for a pre-exisiting user defined route
-    def generate_options_method(path, allow_header)
-      self.class.options(path, {}) do
+    def generate_options_method(path, allow_header, requirements = nil)
```

Comment:
If instead of `requirements` you have `options` (a Hash), you have no work to do here.